packages/tangle/scheduler.go

```diff
@@ -0,0 +1,298 @@
+				issuingTime: message.IssuingTime(),
+				parents:     message.Parents()}
+		})
+	})
+	s.tangle.Solidifier.Events.MessageSolid.Attach(onMessageSolid)
+
+	onMessageBooked := events.NewClosure(func(messageID MessageID) {
+		s.messagesBooked <- messageID
+	})
+	s.tangle.Booker.Events.MessageBooked.Attach(onMessageBooked)
+
+	s.start()
+}
+
+// start starts the scheduler.
+func (s *Scheduler) start() {
+	go func() {
+		s.timeQueue.Start()
+		for {
+			select {
+
+			// read message from the inbox, add them to the timeQueue if their timestamp
+			// is in the future or schedule it if its parents have been booked already.
+			case message := <-s.inbox:
+				if message != nil && message.issuingTime.After(clock.SyncedTime()) {
+					s.timeQueue.Add(message)
+					break
+				}
+				s.trySchedule(message)
+
+			// try to schedule messsage that have been waiting and are now current.
+			case message := <-s.timeQueue.C:
+				s.trySchedule(message)
+
+			// schedule messages that were waiting for their parents to be booked.
+			case messageID := <-s.messagesBooked:
+				for _, child := range s.parentsMap[messageID] {
+					if s.messageReady(child) && !child.scheduled {
+						child.scheduled = true
+						s.schedule(child.ID)
+					}
+				}
+				delete(s.parentsMap, messageID)
+
+			case <-s.close:
+				return
+			}
+		}
+	}()
+}
+
+// Shutdown stops the scheduler and terminates its goroutines and timers.
+func (s *Scheduler) Shutdown() {
+	close(s.close)
+	close(s.inbox)
+
+	s.outboxWorkerPool.ShutdownGracefully()
+	s.timeQueue.Stop()
+}
+
+func (s *Scheduler) schedule(messageID MessageID) {
+	s.outboxWorkerPool.Submit(func() {
+		s.Events.MessageScheduled.Trigger(messageID)
+	})
+}
+
+func (s *Scheduler) trySchedule(message *MessageToSchedule) {
+	if message == nil {
+		return
+	}
+
+	// schedule if all the parents have been booked already.
+	parentsToBook := s.parentsToBook(message)
+	if len(parentsToBook) == 0 {
+		s.schedule(message.ID)
+		return
+	}
+
+	// append the message to the unbooked parent(s) queue(s).
+	for _, parent := range parentsToBook {
+		if _, exist := s.parentsMap[parent]; !exist {
+			s.parentsMap[parent] = make([]*MessageToSchedule, 0)
+		}
+		s.parentsMap[parent] = append(s.parentsMap[parent], message)
+	}
+}
+
+func (s *Scheduler) messageReady(message *MessageToSchedule) (ready bool) {
+	return len(s.parentsToBook(message)) == 0
+}
+
+func (s *Scheduler) parentsToBook(message *MessageToSchedule) (parents MessageIDs) {
+	for _, parentID := range message.parents {
+		s.tangle.Storage.MessageMetadata(parentID).Consume(func(messageMetadata *MessageMetadata) {
+			if !messageMetadata.IsBooked() {
+				parents = append(parents, parentID)
+			}
+		})
+	}
+
+	return parents
+}
+
+// region TimeMessageQueue /////////////////////////////////////////////////////////////////////////////////////////////
+
+// TimeMessageQueue is a time-based ordered queue.
+type TimeMessageQueue struct {
+	list    timeIssuanceSortedList
```

Comment:
what about this list? should this be protected as well?

Reply:
Also managed only by the scheduler

Reply:
I mean, it is internally used by the TimeMessageQueue, which is managed by the scheduler and its access is restricted via a select